File: hello-api/src/main/java/org/dustin/kotlin/hello/api/HelloService.java
package org.dustin.kotlin.hello.api;

import static com.lightbend.lagom.javadsl.api.Service.named;
import static com.lightbend.lagom.javadsl.api.Service.pathCall;
import static com.lightbend.lagom.javadsl.api.Service.topic;

import akka.Done;
import akka.NotUsed;
import com.lightbend.lagom.javadsl.api.Descriptor;
import com.lightbend.lagom.javadsl.api.Service;
import com.lightbend.lagom.javadsl.api.ServiceCall;
import com.lightbend.lagom.javadsl.api.broker.Topic;
import com.lightbend.lagom.javadsl.api.broker.kafka.KafkaProperties;

/**
 * Lagom Service interfaces remains in Java due to default methods not supported in Kotlin.
 *
 * If implement Service in Kotlin, get this runtime error:
 *
 * Caused by: java.lang.IllegalArgumentException: Service.descriptor must be implemented as a default method
 *
 * Workaround: https://github.com/lagom/lagom/issues/1180
 */
public interface HelloService extends Service {

    /**
     * Example: curl http://localhost:9000/api/hello/Alice
     */
    ServiceCall<NotUsed, String> hello(String id);


    /**
     * Example: curl -H "Content-Type: application/json" -X POST -d '{"message": "Hi"}' http://localhost:9000/api/hello/Alice
     */
    ServiceCall<GreetingMessage, Done> useGreeting(String id);


    /**
     * This gets published to Kafka.
     */
    Topic<HelloEvent> helloEvents();

    @Override
    default Descriptor descriptor() {
        // @formatter:off
    return named("hello").withCalls(
        pathCall("/api/hello/:id",  this::hello),
        pathCall("/api/hello/:id", this::useGreeting)
      ).withTopics(
          topic("hello-events", this::helloEvents)
          // Kafka partitions messages, messages within the same partition will
          // be delivered in order, to ensure that all messages for the same user
          // go to the same partition (and hence are delivered in order with respect
          // to that user), we configure a partition key strategy that extracts the
          // name as the partition key.
          .withProperty(KafkaProperties.partitionKeyStrategy(), HelloEvent::getEventName)
        ).withAutoAcl(true);
    // @formatter:on
    }
}

File: hello-api/src/main/kotlin/org/dustin/kotlin/hello/api/HelloService.kt
/**
 * Can't use this until Java interoperability with default methods is supported in Kotlin.
 */

//package org.dustin.kotlin.hello.api
//
//import akka.Done
//import akka.NotUsed
//import com.lightbend.lagom.javadsl.api.Descriptor
//import com.lightbend.lagom.javadsl.api.Service
//import com.lightbend.lagom.javadsl.api.Service.*
//import com.lightbend.lagom.javadsl.api.ServiceCall
//import com.lightbend.lagom.javadsl.api.broker.Topic
//
//
//interface HelloService : Service {
//    /**
//     * Example: curl http://localhost:9000/api/hello/Alice
//     */
//    fun hello(id: String): ServiceCall<NotUsed, String>
//
//
//    /**
//     * Example: curl -H "Content-Type: application/json" -X POST -d '{"message":
//     * "Hi"}' http://localhost:9000/api/hello/Alice
//     */
//    fun useGreeting(id: String): ServiceCall<GreetingMessage, Done>
//
//
//    /**
//     * This gets published to Kafka.
//     */
//    fun helloEvents(): Topic<HelloEvent>
//
//    override fun descriptor(): Descriptor {
//        return named("hello").withCalls(
//                pathCall<NotUsed, String, String>("/api/hello/:id", this::hello),
//                pathCall<GreetingMessage, Done, String>("/api/hello/:id", this::useGreeting)
//        ).withTopics(
//                topic<HelloEvent>("hello-events", this::helloEvents)
//        ).withAutoAcl(true)
//    }
//}